File: emcom/emcom/formatting.py
"""Display helpers for emcom CLI output."""

from __future__ import annotations

from emcom.models import Email, Identity, Thread


def short_id(uuid_str: str) -> str:
    return uuid_str[:8]


def short_date(iso_str: str) -> str:
    """Format ISO timestamp as MM/DD HH:MM:SS."""
    # iso_str is like '2026-03-10T07:09:10...'
    d = iso_str[:19].replace("T", " ")
    # d = '2026-03-10 07:09:10'
    return f"{d[5:7]}/{d[8:10]} {d[11:]}"


def _trunc(s: str, width: int) -> str:
    return s if len(s) <= width else s[: width - 1] + "…"


def format_inbox(emails: list[Email], numbered: bool = False) -> str:
    if not emails:
        return "Inbox is empty."
    id_hdr = "#" if numbered else "ID"
    id_w = 4 if numbered else 8
    lines = [f"{id_hdr:{id_w}}  {'From':12}  {'Subject':30}  {'Date':14}  Tags"]
    lines.append("-" * (id_w + 2 + 12 + 2 + 30 + 2 + 14 + 2 + 4))
    for idx, e in enumerate(emails, 1):
        tags = ", ".join(e.tags) if e.tags else ""
        date = short_date(e.created_at)
        id_col = f"[{idx}]" if numbered else short_id(e.id)
        lines.append(f"{id_col:{id_w}}  {e.sender:12}  {e.subject[:30]:30}  {date:14}  {tags}")
    return "\n".join(lines)


def format_email(email: Email) -> str:
    lines = [
        f"From:    {email.sender}",
        f"To:      {', '.join(email.to)}",
    ]
    if email.cc:
        lines.append(f"CC:      {', '.join(email.cc)}")
    lines.extend([
        f"Subject: {email.subject}",
        f"Date:    {short_date(email.created_at)}",
        f"ID:      {email.id}",
    ])
    if email.tags:
        lines.append(f"Tags:    {', '.join(email.tags)}")
    lines.append("")
    lines.append(email.body)
    return "\n".join(lines)


def format_thread(emails: list[Email]) -> str:
    if not emails:
        return "Thread is empty."
    parts = []
    for i, e in enumerate(emails):
        if i > 0:
            parts.append("---")
        parts.append(format_email(e))
    return "\n".join(parts)


def format_who(identities: list[Identity]) -> str:
    if not identities:
        return "No registered agents."
    desc_width = min(60, max(len("Description"), max(len(i.description) for i in identities)))
    loc_width = min(30, max(len("Location"), max((len(i.location) for i in identities), default=0)))
    lines = [f"{'Name':12}  {'Description':{desc_width}}  {'Location':{loc_width}}  {'Last Seen':14}"]
    lines.append("-" * (12 + 2 + desc_width + 2 + loc_width + 2 + 14))
    for i in identities:
        seen = short_date(i.last_seen)
        desc = _trunc(i.description, desc_width)
        loc = _trunc(i.location, loc_width)
        lines.append(f"{i.name:12}  {desc:{desc_width}}  {loc:{loc_width}}  {seen:14}")
    return "\n".join(lines)


def format_all_mail(emails: list[Email], viewer: str, numbered: bool = False) -> str:
    if not emails:
        return "No emails."
    id_hdr = "#" if numbered else "ID"
    id_w = 4 if numbered else 8
    lines = [f"{id_hdr:{id_w}}  {'':2}  {'From':12}  {'To':12}  {'Subject':25}  {'Date':14}  Tags"]
    lines.append("-" * (id_w + 2 + 2 + 2 + 12 + 2 + 12 + 2 + 25 + 2 + 14 + 2 + 4))
    for idx, e in enumerate(emails, 1):
        direction = ">>" if e.sender == viewer else "<<"
        to = ", ".join(e.to)
        tags = ", ".join(e.tags) if e.tags else ""
        date = short_date(e.created_at)
        id_col = f"[{idx}]" if numbered else short_id(e.id)
        lines.append(f"{id_col:{id_w}}  {direction}  {e.sender:12}  {to[:12]:12}  {e.subject[:25]:25}  {date:14}  {tags}")
    return "\n".join(lines)


def format_sent(emails: list[Email], numbered: bool = False) -> str:
    if not emails:
        return "No sent emails."
    id_hdr = "#" if numbered else "ID"
    id_w = 4 if numbered else 8
    lines = [f"{id_hdr:{id_w}}  {'To':20}  {'Subject':30}  {'Date':14}"]
    lines.append("-" * (id_w + 2 + 20 + 2 + 30 + 2 + 14))
    for idx, e in enumerate(emails, 1):
        to = ", ".join(e.to)
        date = short_date(e.created_at)
        id_col = f"[{idx}]" if numbered else short_id(e.id)
        lines.append(f"{id_col:{id_w}}  {to[:20]:20}  {e.subject[:30]:30}  {date:14}")
    return "\n".join(lines)


def format_threads(threads: list[Thread], numbered: bool = False) -> str:
    if not threads:
        return "No threads."
    id_hdr = "#" if numbered else "Thread ID"
    id_w = 4 if numbered else 8
    lines = [f"{id_hdr:{id_w}}  {'Subject':25}  {'Participants':25}  {'Emails':6}  {'Last Activity':14}"]
    lines.append("-" * (id_w + 2 + 25 + 2 + 25 + 2 + 6 + 2 + 14))
    for idx, t in enumerate(threads, 1):
        parts = ", ".join(t.participants)
        date = short_date(t.last_activity)
        id_col = f"[{idx}]" if numbered else short_id(t.thread_id)
        lines.append(f"{id_col:{id_w}}  {t.subject[:25]:25}  {parts[:25]:25}  {t.email_count:<6}  {date:14}")
    return "\n".join(lines)

File: emcom/emcomcs/Formatting.cs
using System.Text;

namespace Emcom;

public static class Fmt
{
    public static string ShortId(string uuid) => uuid.Length >= 8 ? uuid[..8] : uuid;

    /// <summary>Format ISO timestamp as MM/DD HH:MM:SS (substring, no date parsing).</summary>
    public static string ShortDate(string iso)
    {
        // iso is like '2026-03-10T07:09:10...'
        if (iso.Length < 19) return iso;
        // d = "2026-03-10 07:09:10"
        return $"{iso[5..7]}/{iso[8..10]} {iso[11..19]}";
    }

    public static string Trunc(string s, int width) =>
        s.Length <= width ? s : s[..(width - 1)] + "\u2026";

    // --- List formatters ---

    public static string FormatInbox(List<Email> emails, bool numbered = false)
    {
        if (emails.Count == 0) return "Inbox is empty.";
        var idHdr = numbered ? "#" : "ID";
        var idW = numbered ? 4 : 8;
        var sb = new StringBuilder();
        sb.AppendLine($"{idHdr.PadRight(idW)}  {"From",-12}  {"Subject",-30}  {"Date",-14}  Tags");
        sb.AppendLine(new string('-', idW + 2 + 12 + 2 + 30 + 2 + 14 + 2 + 4));
        for (int i = 0; i < emails.Count; i++)
        {
            var e = emails[i];
            var tags = e.Tags.Count > 0 ? string.Join(", ", e.Tags) : "";
            var date = ShortDate(e.CreatedAt);
            var idCol = numbered ? $"[{i + 1}]" : ShortId(e.Id);
            sb.AppendLine($"{idCol.PadRight(idW)}  {e.SenderName.PadRight(12)}  {Trunc(e.Subject, 30).PadRight(30)}  {date.PadRight(14)}  {tags}");
        }
        return sb.ToString().TrimEnd();
    }

    public static string FormatEmail(Email email)
    {
        var sb = new StringBuilder();
        sb.AppendLine($"From:    {email.SenderName}");
        sb.AppendLine($"To:      {string.Join(", ", email.To)}");
        if (email.Cc.Count > 0)
            sb.AppendLine($"CC:      {string.Join(", ", email.Cc)}");
        sb.AppendLine($"Subject: {email.Subject}");
        sb.AppendLine($"Date:    {ShortDate(email.CreatedAt)}");
        sb.AppendLine($"ID:      {email.Id}");
        if (email.Tags.Count > 0)
            sb.AppendLine($"Tags:    {string.Join(", ", email.Tags)}");
        sb.AppendLine();
        sb.Append(email.Body);
        return sb.ToString();
    }

    public static string FormatThread(List<Email> emails)
    {
        if (emails.Count == 0) return "Thread is empty.";
        var sb = new StringBuilder();
        for (int i = 0; i < emails.Count; i++)
        {
            if (i > 0) sb.AppendLine("---");
            sb.Append(FormatEmail(emails[i]));
            if (i < emails.Count - 1) sb.AppendLine();
        }
        return sb.ToString();
    }

    public static string FormatWho(List<Identity> identities)
    {
        if (identities.Count == 0) return "No registered agents.";
        int descWidth = Math.Min(60, Math.Max("Description".Length, identities.Max(i => i.Description.Length)));
        int locWidth = Math.Min(30, Math.Max("Location".Length, identities.Max(i => i.Location.Length)));
        var sb = new StringBuilder();
        sb.AppendLine($"{"Name",-12}  {"Description".PadRight(descWidth)}  {"Location".PadRight(locWidth)}  {"Last Seen",-14}");
        sb.AppendLine(new string('-', 12 + 2 + descWidth + 2 + locWidth + 2 + 14));
        foreach (var i in identities)
        {
            var seen = ShortDate(i.LastSeen);
            var desc = Trunc(i.Description, descWidth);
            var loc = Trunc(i.Location, locWidth);
            sb.AppendLine($"{i.Name.PadRight(12)}  {desc.PadRight(descWidth)}  {loc.PadRight(locWidth)}  {seen.PadRight(14)}");
        }
        return sb.ToString().TrimEnd();
    }

    public static string FormatAllMail(List<Email> emails, string viewer, bool numbered = false)
    {
        if (emails.Count == 0) return "No emails.";
        var idHdr = numbered ? "#" : "ID";
        var idW = numbered ? 4 : 8;
        var sb = new StringBuilder();
        sb.AppendLine($"{idHdr.PadRight(idW)}  {"",2}  {"From",-12}  {"To",-12}  {"Subject",-25}  {"Date",-14}  Tags");
        sb.AppendLine(new string('-', idW + 2 + 2 + 2 + 12 + 2 + 12 + 2 + 25 + 2 + 14 + 2 + 4));
        for (int i = 0; i < emails.Count; i++)
        {
            var e = emails[i];
            var direction = e.SenderName == viewer ? ">>" : "<<";
            var to = string.Join(", ", e.To);
            var tags = e.Tags.Count > 0 ? string.Join(", ", e.Tags) : "";
            var date = ShortDate(e.CreatedAt);
            var idCol = numbered ? $"[{i + 1}]" : ShortId(e.Id);
            sb.AppendLine($"{idCol.PadRight(idW)}  {direction}  {e.SenderName.PadRight(12)}  {Trunc(to, 12).PadRight(12)}  {Trunc(e.Subject, 25).PadRight(25)}  {date.PadRight(14)}  {tags}");
        }
        return sb.ToString().TrimEnd();
    }

    public static string FormatSent(List<Email> emails, bool numbered = false)
    {
        if (emails.Count == 0) return "No sent emails.";
        var idHdr = numbered ? "#" : "ID";
        var idW = numbered ? 4 : 8;
        var sb = new StringBuilder();
        sb.AppendLine($"{idHdr.PadRight(idW)}  {"To",-20}  {"Subject",-30}  {"Date",-14}");
        sb.AppendLine(new string('-', idW + 2 + 20 + 2 + 30 + 2 + 14));
        for (int i = 0; i < emails.Count; i++)
        {
            var e = emails[i];
            var to = string.Join(", ", e.To);
            var date = ShortDate(e.CreatedAt);
            var idCol = numbered ? $"[{i + 1}]" : ShortId(e.Id);
            sb.AppendLine($"{idCol.PadRight(idW)}  {Trunc(to, 20).PadRight(20)}  {Trunc(e.Subject, 30).PadRight(30)}  {date.PadRight(14)}");
        }
        return sb.ToString().TrimEnd();
    }

    public static string FormatThreads(List<Thread> threads, bool numbered = false)
    {
        if (threads.Count == 0) return "No threads.";
        var idHdr = numbered ? "#" : "Thread ID";
        var idW = numbered ? 4 : 8;
        var sb = new StringBuilder();
        sb.AppendLine($"{idHdr.PadRight(idW)}  {"Subject",-25}  {"Participants",-25}  {"Emails",-6}  {"Last Activity",-14}");
        sb.AppendLine(new string('-', idW + 2 + 25 + 2 + 25 + 2 + 6 + 2 + 14));
        for (int i = 0; i < threads.Count; i++)
        {
            var t = threads[i];
            var parts = string.Join(", ", t.Participants);
            var date = ShortDate(t.LastActivity);
            var idCol = numbered ? $"[{i + 1}]" : ShortId(t.ThreadId);
            sb.AppendLine($"{idCol.PadRight(idW)}  {Trunc(t.Subject, 25).PadRight(25)}  {Trunc(parts, 25).PadRight(25)}  {t.EmailCount.ToString().PadRight(6)}  {date.PadRight(14)}");
        }
        return sb.ToString().TrimEnd();
    }
}
